File: src/main/java/mods/cybercat/gigeresque/common/block/material/MaterialBuilder.java
package mods.cybercat.gigeresque.common.block.material;

import net.minecraft.world.level.material.Material;
import net.minecraft.world.level.material.MaterialColor;
import net.minecraft.world.level.material.PushReaction;

public class MaterialBuilder {
	private PushReaction pistonBehavior;
	private boolean blocksMovement;
	private boolean burnable = false;
	private boolean liquid = false;
	private boolean replaceable = false;
	private boolean solid;
	private MaterialColor color;
	private boolean blocksLight;

	MaterialBuilder(MaterialColor color) {
		pistonBehavior = PushReaction.NORMAL;
		blocksMovement = true;
		solid = true;
		blocksLight = true;
		this.color = color;
	}

	public MaterialBuilder liquid() {
		liquid = true;
		return this;
	}

	public MaterialBuilder notSolid() {
		solid = false;
		return this;
	}

	public MaterialBuilder allowsMovement() {
		blocksMovement = false;
		return this;
	}

	public MaterialBuilder lightPassesThrough() {
		blocksLight = false;
		return this;
	}

	public MaterialBuilder burnable() {
		burnable = true;
		return this;
	}

	public MaterialBuilder replaceable() {
		replaceable = true;
		return this;
	}

	public MaterialBuilder destroyedByPiston() {
		pistonBehavior = PushReaction.DESTROY;
		return this;
	}

	public MaterialBuilder blocksPiston() {
		pistonBehavior = PushReaction.BLOCK;
		return this;
	}

	public Material build() {
		return new Material(color, liquid, solid, blocksMovement, blocksLight, burnable, replaceable, pistonBehavior);
	}
}

File: src/main/kotlin/com/bvanseg/gigeresque/common/block/material/MaterialBuilder.kt
//package com.bvanseg.gigeresque.common.block.material
//
//import net.minecraft.block.MapColor
//import net.minecraft.block.Material
//import net.minecraft.block.piston.PistonBehavior
//
///**
// * @author Boston Vanseghi
// */
//open class MaterialBuilder(color: MapColor) {
//    private var pistonBehavior: PistonBehavior
//    private var blocksMovement: Boolean
//    private var burnable = false
//    private var liquid = false
//    private var replaceable = false
//    private var solid: Boolean
//    private val color: MapColor
//    private var blocksLight: Boolean
//
//    init {
//        pistonBehavior = PistonBehavior.NORMAL
//        blocksMovement = true
//        solid = true
//        blocksLight = true
//        this.color = color
//    }
//
//    fun liquid(): MaterialBuilder {
//        liquid = true
//        return this
//    }
//
//    fun notSolid(): MaterialBuilder {
//        solid = false
//        return this
//    }
//
//    fun allowsMovement(): MaterialBuilder {
//        blocksMovement = false
//        return this
//    }
//
//    fun lightPassesThrough(): MaterialBuilder {
//        blocksLight = false
//        return this
//    }
//
//    fun burnable(): MaterialBuilder {
//        burnable = true
//        return this
//    }
//
//    fun replaceable(): MaterialBuilder {
//        replaceable = true
//        return this
//    }
//
//    fun destroyedByPiston(): MaterialBuilder {
//        pistonBehavior = PistonBehavior.DESTROY
//        return this
//    }
//
//    fun blocksPistons(): MaterialBuilder {
//        pistonBehavior = PistonBehavior.BLOCK
//        return this
//    }
//
//    fun build(): Material = Material(
//        color,
//        liquid, solid, blocksMovement, blocksLight, burnable, replaceable, pistonBehavior
//    )
//}